lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,290 @@
+        real average frequency from the trace.
+        Check that the difference between expected and measured frequencies is
+        no larger than ```freq_margin_pct```.
+        """
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        freqs = self.plat_info['freqs'][cpu]
+        max_freq = max(freqs)
+
+        # Estimate the target frequency, including sugov's margin, and round
+        # into a real OPP
+        target_freq = min(max_freq, max_freq * self.boost / 80)
+        target_freq = list(filter(lambda f: f >= target_freq, freqs))[0]
+
+	# Get the real average frequency
+        avg_freq = self.trace.analysis.frequency.get_average_cpu_frequency(cpu)
+
+        distance = abs(target_freq - avg_freq) * 100 / target_freq
+        res = ResultBundle.from_bool(distance < freq_margin_pct)
+        res.add_metric("target freq", target_freq, 'kHz')
+        res.add_metric("average freq", avg_freq, 'kHz')
+
+        return res
+
+class BoostedMaxFreq(SchedTuneFreqBoost):
+    """
+    A tiny task with 100% boost
+    """
+    boost = 100
+
+class BoostedFreq80(SchedTuneFreqBoost):
+    """
+    A tiny task with 80% boost
+    """
+    boost = 80
+
+class BoostedFreq60(SchedTuneFreqBoost):
+    """
+    A tiny task with 60% boost
+    """
+    boost = 60
+
+class BoostedFreq40(SchedTuneFreqBoost):
+    """
+    A tiny task with 40% boost
+    """
+    boost = 40
+
+class BoostedFreq20(SchedTuneFreqBoost):
+    """
+    A tiny task with 20% boost
+    """
+    boost = 20
+
+class UnboostedFreq(SchedTuneFreqBoost):
+    """
+    A tiny task with no boost
+    """
+    boost = 0
+
+
+class SchedTuneTaskPlacement(SchedTuneTest, abc.ABC):
+    """
+    Abstract class for testing SchedTune's task-placement biasing feature
+
+    Run a tiny rtapp tasks in a cgroup with prefer-idle set, and checks that
+    it runs most of the time on CPUs with enough capacity to meet its boost
+    request.
+    """
+    ftrace_conf = FtraceConf({
+        "events" : ["sched_switch", "sched_wakeup"]
+    }, __qualname__)
+
+    prefer_idle = True
+
+    @classmethod
+    def check_from_target(cls, target):
+        # XXX: better way to do that ?
+        if not target.file_exists('/sys/kernel/debug/tracing/available_events'):
+            raise CannotCreateError("Ftrace not enabled")
+
+        # XXX: better way to do that ?
+        for m in ['cgroups', 'sched']:
+            if not target.has(m):
+                raise CannotCreateError("'{}' devlib module not enabled".format(m))
+
+        for c in ['schedtune']:
+            if c not in target.cgroups.controllers:
+                raise CannotCreateError("'{}' cgroup controller not available".format(c))
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        cpu = plat_info['capacity-classes'][-1][0]
+        rtapp_profile = {}
+        rtapp_profile['rta_stune'] = Periodic(
+            duty_cycle_pct = 1,
+            duration_s = 3,
```

Comment:
Unless I missed something, this profile is very similar to the one in the previous test class - the only diff I see is the duration. Would it make sense to have these merged into a single test class with two test methods?

Reply:
There's another difference, for the frequency tests the task is pinned to a big CPU, and for the task placement it's not ... :/

Reply:
So I did miss something :)